tgui: Window Recycling, Simplified DM API

actioninja opened the pull request:
## About The Pull Request
Avoids closing browser windows in favor of hiding them and later injecting new interfaces into them.

Why?

Setting windows to visible/invisible is much faster than loading new browser windows for a variety of reasons including caching and also just that byond browse calls as slow as hell.

The practical effect of this is that windows now open and fully render *much* faster. Initial open can take 1-3 frames while full render can take 5-10 frames. This is a significant speedup from previously and makes windows feel like the open nearly instantly.
On top of that, the lack of browse calls results in no client hitching on close, and very minimal client hitching on open.
Video demo: https://file.house/EhO6.mp4
The initial open looks as before because that is creating a new window. All subsequent opens use the recycled window and are much faster as a result.

Caveats:
- Windows closed via alt+f4 do not hide and fully destroy, preventing them from being recycled.
- Users of simple/non-fancy mode largely cannot make use of this because we can't override the close button functionality. Auto-close due to distance still will, however.
- Hiding the window potentially has performance concerns and needs some testing

Other additions:
- Client-side window width/height and title.
- Reworked hotkey API, which makes it easier to subscribe to hotkeys and dispatch Redux actions.
- Dev build indicator, which also functions as a KitchenSink toggle.
- Web Storage abstraction in common/storage.js
- Memory of previous window locations, completely client-side window geometry management.
- ui_assets proc (replacement for ui_base_html for dynamically loading assets into tgui.
- Logging Redux middleware, which allows seeing what actions are being dispatched while in `--debug` mode.

## Changelog
:cl:
refactor: tgui windows should load faster after an initial load. Please report any breakages.
/:cl:

actioninja committed:
move subsystem procs to the actual subsystem

actioninja committed:
wip

actioninja committed:
first functional version

actioninja committed:
properly working + standby

actioninja committed:
fixes

actioninja committed:
remove some debug prints

actioninja committed:
deprecated var removal

stylemistake commented on tgui/packages/tgui/layouts/Window.js:
```diff
@@ -26,37 +27,45 @@ export class Window extends Component {
       debugLayout,
+      standby,
     } = useBackend(this.context);
+    const { store } = this.context;
```
Use this instead, since this is more in line with the current React Redux API:
```
const dispatch = useDispatch(this.context);
```

stylemistake commented on tgui/packages/tgui/layouts/Window.js:
```diff
@@ -26,37 +27,45 @@ export class Window extends Component {
-            <div className="Window__dimmer" />
-          )}
-        </div>
+        {!standby ? (
```
Preferrably should be `!standby && ()`, since it's shorter (and booleans do not render).

stylemistake commented on tgui/packages/tgui/standby.js:
```diff
@@ -0,0 +1,24 @@
+export const enterStandby = () => ({
+  type: 'standby',
+});
+
+export const standbyReducer = (state, action) => {
+  const { type } = action;
+
+  if (type === 'standby') {
+    return {
+      ...state,
+      data: {},
+      shared: {},
+      config: {
+        ...state.config,
+        title: '',
+        status: 1,
+        interface: '',
+      },
+      standby: true,
+    };
+  }
+
+  return state;
+};
```
This would belong better to the backend module, since it's the part of the backend, and you DO access `standby` var via `useBackend()`, so it really makes sense to keep it all in one module.

Join reducers as well, and action type should also be called `backend/enterStandby`.

stylemistake commented on tgui/packages/tgui/index.js:
```diff
@@ -136,6 +137,20 @@ const setupApp = () => {
     store.dispatch(backendUpdate(state));
   };
 
+  window.reinit = newRef => {
+    logger.log("Reinitializing to: " + newRef, store.getState());
+    window.__ref__ = newRef;
+    callByond('', {
+      src: window.__ref__,
+      action: 'tgui:initialize',
+    });
+  };
```
This is a source of a lag, because it becomes a long chain of byond calls.

```
window.reinit -> act('tgui:initialize') -> window.update
```

This could just be a single update callback:

```js
window.update = () => {
  ...
  if (state.standby) {
    logger.log("Reinitializing to: " + state.config.ref);
    window.__ref__ = state.config.ref;
  }
  ...
};
```

If window is recycled, just send a full proper update. I guarantee you will save a bunch of frames.

stylemistake commented on code/modules/tgui/tgui.dm:
```diff
@@ -96,35 +96,45 @@
+		free_windows -= window_id
+		winset(user.client, window_id, "size=[width]x[height];is-visible=true")
+		initialized = TRUE
+		user << output(url_encode(ref(src)), "[window_id].browser:reinit")
```
Send a full update here instead of `reinit`.

actioninja committed:
refactor ui state

actioninja committed:
some fixes

actioninja committed:
? to &&

actioninja committed:
dispatch

actioninja committed:
move standby

actioninja committed:
handle leaving standby in update

actioninja committed:
lint fixes

actioninja committed:
move initial update to update

actioninja committed:
move this out of the constructor

stylemistake committed:
Fix a few bugs, add a dev build indicator

stylemistake committed:
Rename into suspend, more bugs

stylemistake commented:
spritesheets that are added via ui_base_html are not loading

stylemistake committed:
Persist window geometries

stylemistake commented:
So uh, added storing/recalling of window positions, and support for width/height on Window, but few things need to be addressed:

- Currently, localStorage will fill with garbage over time. I'm thinking to add a list of recently used UIs, and limit it to 100 or so. Frequently used UIs will be remembered, while random UIs will be forgotten, and it will be self-cleaning as well.
- When window is resized, it shows a white trail for a few frames, reused windows flicker a bit. No idea how to fix this.

stylemistake committed:
Check this out, MORE CODE!!!

- Dynamic loading of assets via ui_assets proc
- Proper storing/recalling of geometries, automatic centering
- Misc

stylemistake committed:
Reworked hotkeys, fixed complicated bugs with window recycling

stylemistake committed:
Convert Biogenerator's ui_base_html into ui_assets

- Added license to assets.js and debug.js

stylemistake committed:
Remove debug prints

stylemistake commented:
## New Additions

- Client-side window width/height and title.
- Reworked hotkey API, which makes it easier to subscribe to hotkeys and dispatch Redux actions.
- Dev build indicator, which also functions as a KitchenSink toggle.
- Web Storage abstraction in `common/storage.js`
- Memory of previous window locations, completely client-side window geometry management.
- `ui_assets` proc (replacement for `ui_base_html` for dynamically loading assets into tgui.
- Logging Redux middleware, which allows seeing what actions are being dispatched while in `--debug` mode.
- Fixed a crapton of bugs with window recycling.

[WebM speed demonstration](https://cdn.discordapp.com/attachments/484170915253321734/709137368460034138/output.webm).

![image](https://user-images.githubusercontent.com/1516236/81516458-99751a80-9340-11ea-9c26-aee79eb31472.png)

## TODOs

- Start browser window as hidden initially, to hide various startup jank.
- Recount frames to update the benchmark table.

stylemistake committed:
Fix vending machines, react to changes of "fancy"

stylemistake committed:
Rebuild tgui

stylemistake committed:
Fix suspend code and few runtimes where user.client is null

stylemistake committed:
Testmerge stuff

stylemistake committed:
More testmerge stuff

stylemistake committed:
Various improvements

- Improve error reporting
- Refactor backend redux module to be contained within the
  state.backend namespace.
- `common/perf` module for measuring performance of the code.

stylemistake committed:
Add missing selector to useSharedState

stylemistake committed:
Release held keys at a more appropriate point in time

stylemistake committed:
More robust logging (doesn't require a correct src anymore)

stylemistake committed:
Big brain

Vanadiom commented:
Another moment of whack

Vanadiom commented:
Probs on the bluescreen tho, whoever made it.

stylemistake committed:
Fix the null window_id problem by skipping the code in close proc

actioninja committed:
fixes all computers incorrectly receiving the physical state checker

actioninja committed:
fixes all computers incorrectly having not incapacited state

stylemistake committed:
Force close windows that can't be suspended

stylemistake committed:
Huge refactor to simplify ui_interact

stylemistake committed:
Simple tgui assets, tgui v3.1

stylemistake committed:
Remove world.log

stylemistake commented:
It is done, just needs some more testing.

actioninja commented:
dnm because there is fairly serious of issues with this that have not been resolved at the moment.

actioninja commented:
I don't know if this is going to feasibly work as intended with byond's current functionality unfortunately, I am working on pulling out the changes unrelated to window recycling itself.

Dennok commented:
sorrow
